llrt_modules/src/modules/fs/file_handle.rs

```diff
@@ -0,0 +1,559 @@
+use crate::buffer::Buffer;
+use crate::utils::array_buffer::ArrayBufferView;
+
+const DEFAULT_BUFFER_SIZE: usize = 16384;
+const DEFAULT_ENCODING: &str = "utf8";
+
+#[rquickjs::class]
+#[derive(rquickjs::class::Trace)]
+pub struct FileHandle {
+    #[qjs(skip_trace)]
+    file: Option<File>,
+    #[qjs(skip_trace)]
+    path: PathBuf,
+}
+
+impl FileHandle {
+    pub fn new(file: File, path: PathBuf) -> Self {
+        Self {
+            file: Some(file),
+            path,
+        }
+    }
+
+    fn file(&self, ctx: &Ctx<'_>) -> Result<&File> {
+        self.file.as_ref().or_throw_msg(ctx, "FileHandle is closed")
+    }
+
+    fn file_mut(&mut self, ctx: &Ctx<'_>) -> Result<&mut File> {
+        self.file.as_mut().or_throw_msg(ctx, "FileHandle is closed")
+    }
+}
+
+#[rquickjs::methods(rename_all = "camelCase")]
+impl FileHandle {
+    async fn chmod(&self, ctx: Ctx<'_>, mode: u32) -> Result<()> {
+        #[cfg(unix)]
+        {
+            use std::os::unix::fs::PermissionsExt;
+            let perm = std::fs::Permissions::from_mode(mode);
+            self.file(&ctx)?
+                .set_permissions(perm)
+                .await
+                .or_throw_msg(&ctx, "Can't modify file permissions")?;
+        }
+        Ok(())
+    }
+
+    async fn chown(&self, ctx: Ctx<'_>, uid: u32, gid: u32) -> Result<()> {
+        #[cfg(unix)]
+        {
+            let path = self.path.clone();
+            task::spawn_blocking(move || std::os::unix::fs::chown(&path, Some(uid), Some(gid)))
+                .await
+                .or_throw(&ctx)?
+                .or_throw_msg(&ctx, "Can't modify file owner")?;
+        }
+        Ok(())
+    }
+
+    async fn close(&mut self) {
+        if let Some(file) = self.file.take() {
+            drop(file.into_std().await);
+        }
+    }
+
+    async fn datasync(&self, ctx: Ctx<'_>) -> Result<()> {
+        self.file(&ctx)?
+            .sync_data()
+            .await
+            .or_throw_msg(&ctx, "Can't sync file data")?;
+        Ok(())
+    }
+
+    #[qjs(get)]
+    async fn fd(&self, ctx: Ctx<'_>) -> Result<isize> {
+        #[cfg(unix)]
+        {
+            use std::os::fd::AsRawFd;
+            self.file(&ctx)?.as_raw_fd().try_into().or_throw(&ctx)
+        }
+        #[cfg(windows)]
+        {
+            use std::os::windows::io::AsRawHandle;
+            let handle = self.file(&ctx)?.as_raw_handle();
+            Ok(handle as isize)
+        }
+        #[cfg(not(any(unix, windows)))]
+        {
+            Ok(0)
+        }
+    }
+
+    async fn read<'js>(
+        &mut self,
+        ctx: Ctx<'js>,
+        buffer_or_options: Opt<Either<ArrayBufferView<'js>, ReadOptions<'js>>>,
+        options_or_offset: Opt<Either<ReadOptions<'js>, usize>>,
+        length: Opt<usize>,
+        // Not supporting position for now since it is not available in tokio.
+        // See https://github.com/tokio-rs/tokio/issues/699
+        // position: Opt<Either<isize, Null>>,
```

Comment:
The `File` struct already contains a cursor if I'm not mistaken? Does node really guarantee atomic offsets reads? For instance if we do Promise.all with 5 file reads at an offset?
I think it's essential. It's also fine to spawn a task around std file if that makes things a bit simpler to work with.

Reply:
It does contain the cursor but it's an await to get it AFAIK.
It does guarantee atomic read AFAIK since it uses `pread` which doesnt modify the read cursor.
Ok ok, will implement!